### ISSUE TITLE
CA2000 - False positive with WebClient.DownloadStringAsync

### ISSUE THREAD

**Issue by elyoh** [dotnet/roslyn-analyzers#2696]
#### Analyzer package

[Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

#### Package Version

v2.9.3 (Latest)
2.9.4-beta1.final

#### Diagnostic ID

[CA2000](https://docs.microsoft.com/en-gb/visualstudio/code-quality/ca2000-dispose-objects-before-losing-scope)

#### Repro steps

The following code causes a CA2000 warning:

```csharp
var uri = new Uri("dummy url");
var webClient = new WebClient();

webClient.DownloadStringCompleted += (sender, eventArgs) =>
{
        // DownloadStringAsync completed so dispose the WebClient.
        (sender as WebClient)?.Dispose();
};

try
{
        webClient.DownloadStringAsync(uri);
}
catch
{
        // DownloadStringAsync can throw System.Net.WebException or ArgumentNullException
        // without triggering DownloadStringCompleted so dispose the WebClient in these 
        // cases.
        webClient.Dispose();
}
```

#### Expected behavior

CA2000 should not be raised.  The WebClient object is disposed after downloading the string or if DownloadStringAsync throws an exception.

The same issue occurs if DownloadFileAsync or DownloadDataAsync are used.

#### Actual behavior

CA2000 is raised.

**Comment by elyoh** [comment]
This also raises IDE0067 in VisualStudio 16.2.0.

**Comment by mavasani** [comment]
@elyoh I believe you will need to add an in-source suppression for this case. There are bunch of API semantics that are being relied upon here to ensure dispose on all paths:

1. `DownloadStringCompleted` event is always raised if `DownloadStringAsync` has been invoked.
2. The `sender` argument for the raised event is the same `webClient` instance on which event handler was added.

Neither of the above facts can be determined by analysis. CA2000 is aggressive by default and is not conservative in presence of external API invocations on disposable instance.
